File: Implementations/Swift/Sources/AstrolabeProtocol/Negotiation/RuntimeProtocolVersion.swift
//
//  RuntimeProtocolVersion.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/16.
//

public struct RuntimeProtocolVersion: Codable, Comparable, Sendable {
    /// Incompatible contract generation.
    public let major: UInt16

    /// Backward-compatible contract revision.
    public let minor: UInt16

    public init(major: UInt16, minor: UInt16) {
        self.major = major
        self.minor = minor
    }

    public static let v2 = RuntimeProtocolVersion(major: 2, minor: 0)

    public static func < (lhs: RuntimeProtocolVersion, rhs: RuntimeProtocolVersion) -> Bool {
        lhs.major == rhs.major ? lhs.minor < rhs.minor : lhs.major < rhs.major
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            major: try container.decode(UInt16.self, forKey: .major),
            minor: try container.decode(UInt16.self, forKey: .minor)
        )
    }

    private enum CodingKeys: String, CodingKey {
        case major
        case minor
    }
}

public struct RuntimeProtocolRange: Codable, Equatable, Sendable {
    /// Oldest protocol version accepted by the peer.
    public let minimum: RuntimeProtocolVersion

    /// Newest protocol version accepted by the peer.
    public let maximum: RuntimeProtocolVersion

    public init(
        minimum: RuntimeProtocolVersion,
        maximum: RuntimeProtocolVersion
    ) throws {
        guard minimum.major == maximum.major else {
            throw RuntimeProtocolVersionError.crossMajorRange(minimum: minimum, maximum: maximum)
        }
        guard minimum <= maximum else {
            throw RuntimeProtocolVersionError.invalidRange(minimum: minimum, maximum: maximum)
        }
        self.minimum = minimum
        self.maximum = maximum
    }

    public static let v2 = RuntimeProtocolRange(uncheckedMinimum: .v2, maximum: .v2)

    public func contains(_ version: RuntimeProtocolVersion) -> Bool {
        minimum <= version && version <= maximum
    }

    public func highestCommonVersion(with other: RuntimeProtocolRange) -> RuntimeProtocolVersion? {
        let lowerBound = max(minimum, other.minimum)
        let upperBound = min(maximum, other.maximum)
        return lowerBound <= upperBound ? upperBound : nil
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        try self.init(
            minimum: container.decode(RuntimeProtocolVersion.self, forKey: .minimum),
            maximum: container.decode(RuntimeProtocolVersion.self, forKey: .maximum)
        )
    }

    private init(
        uncheckedMinimum minimum: RuntimeProtocolVersion,
        maximum: RuntimeProtocolVersion
    ) {
        self.minimum = minimum
        self.maximum = maximum
    }

    private enum CodingKeys: String, CodingKey {
        case minimum
        case maximum
    }
}

public enum RuntimeProtocolVersionError: Error, Equatable, Sendable {
    case unsupportedVersion(RuntimeProtocolVersion)
    case invalidRange(minimum: RuntimeProtocolVersion, maximum: RuntimeProtocolVersion)
    case crossMajorRange(minimum: RuntimeProtocolVersion, maximum: RuntimeProtocolVersion)
}

File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/negotiation/RuntimeProtocolVersion.kt
//
//  RuntimeProtocolVersion.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable

/** Wire protocol version carried by each request and response. */
@Serializable
public data class RuntimeProtocolVersion(
    /** Protocol major version. */
    public val major: Int,
    /** Protocol minor version. */
    public val minor: Int
) : Comparable<RuntimeProtocolVersion> {
    init {
        require(major in 0..MAXIMUM_VERSION_COMPONENT) { "Protocol major version is outside UInt16 range" }
        require(minor in 0..MAXIMUM_VERSION_COMPONENT) { "Protocol minor version is outside UInt16 range" }
    }

    /** Whether this value is the only protocol version implemented by this package. */
    public val isSupported: Boolean
        get() = this == V2

    override fun compareTo(other: RuntimeProtocolVersion): Int =
        compareValuesBy(this, other, RuntimeProtocolVersion::major, RuntimeProtocolVersion::minor)

    public companion object {
        /** Astrolabe Wire Protocol 2.0. */
        public val V2: RuntimeProtocolVersion = RuntimeProtocolVersion(major = 2, minor = 0)
    }
}

/** Inclusive protocol version interval supported by one peer. */
@Serializable
public data class RuntimeProtocolRange(
    /** Oldest protocol version accepted by the peer. */
    public val minimum: RuntimeProtocolVersion,
    /** Newest protocol version accepted by the peer. */
    public val maximum: RuntimeProtocolVersion
) {
    init {
        require(minimum.major == maximum.major) { "Protocol range cannot span major versions" }
        require(minimum <= maximum) { "Protocol range minimum cannot exceed maximum" }
    }

    /** Returns whether this range includes the supplied version. */
    public operator fun contains(version: RuntimeProtocolVersion): Boolean = version in minimum..maximum

    /** Returns the newest version accepted by both ranges, or null when they do not overlap. */
    public fun highestCommonVersion(other: RuntimeProtocolRange): RuntimeProtocolVersion? {
        val lowerBound = maxOf(minimum, other.minimum)
        val upperBound = minOf(maximum, other.maximum)
        return upperBound.takeIf { lowerBound <= it }
    }

    public companion object {
        /** Range containing Astrolabe Wire Protocol 2.0 only. */
        public val V2: RuntimeProtocolRange = RuntimeProtocolRange(
            minimum = RuntimeProtocolVersion.V2,
            maximum = RuntimeProtocolVersion.V2
        )
    }
}

private const val MAXIMUM_VERSION_COMPONENT: Int = 65_535
